code/game/sound.dm

```diff
@@ -1,104 +1,140 @@
-
-	soundin = get_sfx(soundin) // same sound for everyone
 
+/proc/playsound(atom/source, input, vol as num, vary, extrarange as num, falloff, frequency = null, channel = 0, pressure_affected = TRUE, ignore_walls = TRUE, do_owner)
 	if(isarea(source))
-		CRASH("[source] is an area and is trying to make the sound: [soundin]")
+		CRASH("playsound(): source is an area")
+		return
 
-	var/frequency = GET_RANDOM_FREQ // Same frequency for everybody
 	var/turf/turf_source = get_turf(source)
-	if(!turf_source) return
+
+	if (!turf_source)
+		return
+
+	//allocate a channel if necessary now so its the same for everyone
+	channel = channel || open_sound_channel()
+
  	// Looping through the player list has the added bonus of working for mobs inside containers
-	var/mob/M
-	var/turf/T
-	for(var/i in GLOB.player_list)
-		M = i
-		if(!istype(M) || !M.client) continue
-		if(get_dist(M, turf_source) <= sound_range)
-			T = get_turf(M)
-			if(T && T.z == turf_source.z) M.playsound_local(turf_source, soundin, vol, vary, frequency, falloff, is_global)
-
-/mob/proc/playsound_local(turf/turf_source, soundin, vol, vary, frequency, var/falloff, is_global)
-	if(!client || ear_deaf > 0)	return FALSE
-	soundin = get_sfx(soundin)
+	var/maxdistance = (world.view + extrarange)
+	var/z = turf_source.z
+	var/list/listeners = SSmobs.clients_by_zlevel[z]
+	if(!ignore_walls) //these sounds don't carry through walls
+		listeners = listeners & hearers(maxdistance, turf_source)
+
+	for(var/P in listeners)
+		var/mob/M = P
+		
+		if(get_dist(M, turf_source) <= maxdistance)
+			sound_or_datum(M, turf_source, source, input, do_owner, vol, vary, frequency, falloff, channel, pressure_affected)
+	
+	for(var/P in SSmobs.dead_players_by_zlevel[z])
+		var/mob/M = P
+		if(get_dist(M, turf_source) <= maxdistance)
+			sound_or_datum(M, turf_source, source, input, do_owner, vol, vary, frequency, falloff, channel, pressure_affected)
+			
+/proc/sound_or_datum(mob/receiver, turf/turf_source, atom/source, input, datum/do_owner, vol as num, vary, frequency, falloff, channel = 0, pressure_affected = TRUE)
+	if(istype(input, /datum/outputs))
+		var/last_played_time
+		if(do_owner)
+			last_played_time = do_owner.datum_outputs[input]
+		else
+			last_played_time = source.datum_outputs[input]
+		var/datum/outputs/O = input
+		if(O.send_info(receiver, turf_source, vol, vary, frequency, falloff, channel, pressure_affected, last_played_time))
+			if(do_owner)
+				do_owner.datum_outputs[input] = world.time
+			else
+				source.datum_outputs[input] = world.time
+	else
+		var/sound/S = sound(get_sfx(input))
+		receiver.playsound_local(turf_source, input, vol, vary, frequency, falloff, channel, pressure_affected, S)
+
+//kept for legacy support and uploaded admin sounds
+/mob/proc/playsound_local(turf/turf_source, soundin, vol as num, vary, frequency, falloff, channel = 0, pressure_affected = TRUE, sound/S)
```

Comment:
Since this is legacy and only used in 4 places, I think it would be safe to remove

Reply:
I feel those places need more work to get this removed. For example the grenade sound shouts (in human) "grenade out". This should only be heard / understood by other humans and might require an extra proc.

For now i think its best to leave it in until we can fully replace it?

Reply:
What does it do differently than the new one?